core/src/main/java/org/apache/iceberg/Snapshots.java

```diff
@@ -0,0 +1,226 @@
+    private final long snapshotId;
+    private final int startFileIndex;
+    private final int endFileIndex;
+    private final long sizeInBytes;
+    private final CloseableIterable<FileScanTask> tasks;
+    private final boolean lastIndexOfSnapshot;
+
+    private MicroBatch(long snapshotId, int startFileIndex, int endFileIndex, long sizeInBytes,
+               CloseableIterable<FileScanTask> tasks, boolean lastIndexOfSnapshot) {
+      this.snapshotId = snapshotId;
+      this.startFileIndex = startFileIndex;
+      this.endFileIndex = endFileIndex;
+      this.sizeInBytes = sizeInBytes;
+      this.tasks = tasks;
+      this.lastIndexOfSnapshot = lastIndexOfSnapshot;
+    }
+
+    public long snapshotId() {
+      return snapshotId;
+    }
+
+    public int startFileIndex() {
+      return startFileIndex;
+    }
+
+    public int endFileIndex() {
+      return endFileIndex;
+    }
+
+    public long sizeInBytes() {
+      return sizeInBytes;
+    }
+
+    public CloseableIterable<FileScanTask> tasks() {
+      return tasks;
+    }
+
+    public boolean lastIndexOfSnapshot() {
+      return lastIndexOfSnapshot;
+    }
+  }
+
+  public static class MicroBatchBuilder {
+    private final Snapshot snapshot;
+    private final FileIO io;
+    private Expression rowFilter;
+    private boolean caseSensitive;
+    private Map<Integer, PartitionSpec> specsById;
+
+    private MicroBatchBuilder(Snapshot snapshot, FileIO io) {
+      this.snapshot = snapshot;
+      this.io = io;
+      this.rowFilter = Expressions.alwaysTrue();
+      this.caseSensitive = true;
+    }
+
+    public MicroBatchBuilder caseSensitive(boolean sensitive) {
+      this.caseSensitive = sensitive;
+      return this;
+    }
+
+    public MicroBatchBuilder filter(Expression newRowFilter) {
+      this.rowFilter = newRowFilter;
+      return this;
+    }
+
+    public MicroBatchBuilder specsById(Map<Integer, PartitionSpec> specs) {
+      this.specsById = specs;
+      return this;
+    }
+
+    public MicroBatch generate(int startFileIndex, long targetSizeInBytes, boolean isStarting) {
+      List<ManifestFile> manifests = isStarting ? snapshot.dataManifests() :
+          snapshot.dataManifests().stream().filter(m -> m.snapshotId().equals(snapshot.snapshotId()))
+              .collect(Collectors.toList());
+
+      List<Pair<ManifestFile, Integer>> manifestIndexes = indexManifests(manifests);
+      List<Pair<ManifestFile, Integer>> skippedManifestIndexes = skipManifests(manifestIndexes, startFileIndex);
+
+      return generateMicroBatch(skippedManifestIndexes, startFileIndex, targetSizeInBytes, isStarting);
+    }
+
+    private List<Pair<ManifestFile, Integer>> indexManifests(List<ManifestFile> manifestFiles) {
+      int currentFileIndex = 0;
+      List<Pair<ManifestFile, Integer>> manifestIndexes = Lists.newArrayList();
+
+      for (ManifestFile manifest : manifestFiles) {
+        int filesCount = manifest.addedFilesCount() + manifest.existingFilesCount();
```

Comment:
Nit: `filesCount` is calculated here, but has nothing to do with the next line. It would be easier to read if this were moved after the pair is created, so that it is clear that `currentFileIndex` is the index at the start of the loop. Then you also don't really need to have a separate variable, too:

```java
for (ManifestFile manifest : manifestFiles) {
  manifestIndexes.add(Pair.of(manifest, currentFileIndex));
  currentFileIndex += manifest.addedFilesCount() + manifest.existingFilesCount();
}
```

Reply:
Changed.